modules/homa-lite/src/lib.rs

```diff
@@ -256,5 +276,304 @@ pub mod module {
+				.saturating_sub(request_extra_fee);
+			staking_amount = fee_deducted_percentage.mul(staking_amount);
+
+			// Transfer the reduced staking currency from Minter to Redeemer
+			T::Currency::transfer(T::StakingCurrencyId::get(), minter, redeemer, staking_amount)?;
+
+			new_balances.push((redeemer.clone(), new_amount, request_extra_fee));
+			Self::deposit_event(Event::<T>::Redeemed(
+				redeemer.clone(),
+				staking_amount,
+				liquid_amount_can_be_redeemed,
+			));
+
+			Ok(())
+		}
+
+		/// Mint some Liquid currency, by locking up the given amount of Staking currency.
+		/// The redeem requests given in `requests` are prioritized to be matched. All other redeem
+		/// requests are matched after. The remaining amount is minted through Staking on the
+		/// Relaychain (via XCM).
+		///
+		/// Parameters:
+		/// - `amount`: The amount of Staking currency to be exchanged.
+		/// - `requests`: The redeem requests that are prioritized to match.
+		fn do_mint_with_requests(
+			minter: &T::AccountId,
+			amount: Balance,
+			requests: Vec<T::AccountId>,
+		) -> DispatchResult {
+			// Ensure the amount is above the minimum, after the MintFee is deducted.
+			ensure!(
+				amount > T::MinimumMintThreshold::get().saturating_add(T::MintFee::get()),
+				Error::<T>::MintAmountBelowMinimumThreshold
+			);
+
+			let staking_currency = T::StakingCurrencyId::get();
+
+			// ensure the user has enough funds on their account.
+			T::Currency::ensure_can_withdraw(staking_currency, &minter, amount)?;
+
+			// Attempt to match redeem requests if there are any.
+			let total_liquid_to_mint = Self::convert_staking_to_liquid(amount)?;
+
+			// The amount of liquid currency to be redeemed for the mint reuqest.
+			let mut liquid_remaining = total_liquid_to_mint;
+
+			// New balances after redeem requests are fullfilled.
+			let mut new_balances: Vec<(T::AccountId, Balance, Permill)> = vec![];
+
+			// Iterate through the prioritized requests first
+			for redeemer in requests {
+				// If all the currencies are minted, return.
+				if liquid_remaining.is_zero() {
+					break;
+				}
+
+				// Check if the redeem request exists
+				if let Some((request_amount, extra_fee)) = Self::redeem_requests(&redeemer) {
+					Self::match_mint_with_redeem_request(
+						&minter,
+						&redeemer,
+						request_amount,
+						extra_fee,
+						&mut liquid_remaining,
+						&mut new_balances,
+					)?;
+				}
+			}
+
+			// Update storage to the new balances. Remove Redeem requests that have been filled.
+			for (redeemer, new_balance, extra_fee) in &new_balances {
+				if new_balance.is_zero() {
+					RedeemRequests::<T>::remove(&redeemer);
+				} else {
+					RedeemRequests::<T>::insert(&redeemer, (new_balance, extra_fee));
+				}
+			}
+			// Redeem request storage has now been updated.
+			new_balances.clear();
+
+			// Iterate all remaining redeem requests now.
+			for (redeemer, (request_amount, extra_fee)) in RedeemRequests::<T>::iter() {
```

Comment:
add a const to limit maximum iterations to 20, this also helps to provide an upper bound for benchmarking